File: jump-game-ii.js
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {number[]} nums
 * @return {number}
 */
const jump = (nums) => {
    if (!nums || !nums.length || nums.length == 1) return 0;
    
    let scope = 0;
    let farthest = 0;
    let jumps = 0;
    
    for (let i = 0; i < nums.length; i++) {
        farthest = Math.max(farthest, i + nums[i]);
        if (farthest >= nums.length - 1) return jumps + 1;
        if (i == scope) {
            jumps++;
            scope = farthest;
        }
    }
};

File: JumpGame-II.py
"""
Time Complexity : O(n) where n is the number of elements in nums array
Space Complexity : O(1)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
"""
class Solution:
    def jump(self, nums: List[int]) -> int:
        if len(nums) == 0 or len(nums) < 2:
            return 0
        n = len(nums)
        # Set current and farthest pointers as first element
        curr = farthest = nums[0]
        jumps = 1
        # Traverse through the array and update farthest only when we find anything
        # greater than our current farthest value. Whenever we reach the decided
        # jump value we reset the current pointer to our farthest and increment the 
        # jumps
        for i in range(n):
            farthest = max(farthest, nums[i] + i)
            if i < n - 1 and curr == i:
                curr = farthest
                jumps += 1
        return jumps